File: tools/mm2_event_lang/cfg_lift.py
"""Lift lowered opcode streams to readable structured statements."""

from __future__ import annotations

from decode_event import decode_u16_arg

from .ast import Expr, LoweredOp, Stmt
from .semantic_tables import (
    COND_SET_OPS,
    TRANSITION_NAMES,
)
from .token_util import skip_n_tokens


def _index_at_offset(ops: list[LoweredOp], off: int) -> int:
    for i, op in enumerate(ops):
        if op.off >= off:
            return i
    return len(ops)


def _op_to_cond_expr(op: LoweredOp, strings: list[str], items: list[str]) -> Expr:
    o, a = op.op, op.args
    if o == 0x09:
        return Expr.yes_no(0)
    if o == 0x0A:
        return Expr.yes_no(1)
    if o == 0x16 and len(a) >= 2:
        return Expr.monster_present(a[0], a[1])
    if o == 0x1B and a:
        return Expr.threshold(a[0])
    if o == 0x22 and len(a) >= 2:
        return Expr.era_in(a[0], a[1])
    if o == 0x23 and len(a) >= 2:
        return Expr.day_in(a[0], a[1])
    if o == 0x24:
        v = decode_u16_arg(o, a)
        return Expr.gold_at_least(v if v is not None else 0)
    if o == 0x25:
        v = decode_u16_arg(o, a)
        return Expr.code16(v if v is not None else 0)
    if o == 0x28 and len(a) >= 2:
        return Expr.has_item_id(a[1], a[0])
    if o == 0x2D and len(a) >= 2:
        return Expr.class_field(a[0], a[1])
    if o == 0x17 and len(a) >= 2:
        return Expr("load_var8", {"group": a[0], "index": a[1]})
    if o == 0x30:
        decoded = "".join(
            chr((0x11A - b) & 0xFF) if 0x20 <= ((0x11A - b) & 0xFF) <= 0x7E else "?"
            for b in a
        ).rstrip()
        return Expr.answer_eq(decoded)
    return Expr.raw_op(o, a)


def _op_to_stmt(op: LoweredOp, strings: list[str], items: list[str]) -> Stmt | None:
    o, a = op.op, op.args
    if o in COND_SET_OPS:
        return None
    if o in (0x10, 0x11, 0x2B):
        return None
    if o in (0x01, 0x02, 0x03, 0x04, 0x05, 0x06) and a:
        variants = {0x01: "basic", 0x02: "block", 0x03: "", 0x04: "door", 0x05: "popup_a", 0x06: "popup_b"}
        return Stmt.say(variants[o], a[0])
    if o == 0x07:
        return Stmt.wait("space")
    if o == 0x08:
        return Stmt.wait("key")
    if o == 0x09:
        return Stmt.ask_yes_no(0)
    if o == 0x0A:
        return Stmt.ask_yes_no(1)
    if o == 0x0B and len(a) >= 2:
        return Stmt.service_title(a[0], a[1])
    if o == 0x0C and len(a) >= 2:
        label = TRANSITION_NAMES.get((a[0], a[1]), "")
        return Stmt.go_to(a[0], a[1], label)
    if o == 0x0D and a:
        return Stmt.engine_call(a[0])
    if o == 0x0E and a:
        return Stmt.selector(a[0])
    if o == 0x0F:
        return Stmt.end()
    if o == 0x12 and len(a) >= 12:
        return Stmt.fight(a[:10], a[10:12])
    if o == 0x13 and len(a) >= 10:
        return Stmt.fight_b(a[:10])
    if o == 0x14:
        return Stmt.clear_tile_event()
    if o == 0x15 and len(a) >= 3:
        return Stmt.apply_party(a[0], a[1], a[2])
    if o == 0x18 and len(a) >= 4:
        return Stmt.apply_party_masked(a[0], a[1], a[2], a[3])
    if o == 0x17 and len(a) >= 2:
        return Stmt.load_var8(a[0], a[1])
    if o == 0x1A and len(a) >= 2:
        return Stmt.store_var8(a[0], a[1])
    if o == 0x1E and a:
        return Stmt.delay(a[0])
    if o == 0x21 and len(a) >= 3:
        y, x = (a[0] >> 4) & 0xF, a[0] & 0xF
        return Stmt.set_tile(y, x, a[1], a[2])
    if o == 0x29:
        return Stmt.abort()
    if o == 0x2F:
        return Stmt.clear_input()
    return Stmt.raw_op(o, a)


def lift_range(
    ops: list[LoweredOp],
    seg: bytes,
    start: int,
    end: int,
    strings: list[str],
    items: list[str],
) -> list[Stmt]:
    body: list[Stmt] = []
    i = start
    pending_cond: Expr | None = None

    while i < end:
        op = ops[i]

        if op.op in COND_SET_OPS:
            pending_cond = _op_to_cond_expr(op, strings, items)
            i += 1
            continue

        if op.op in (0x10, 0x11) and op.args:
            n = op.args[0]
            skip_start = op.off + 2
            target_off = skip_n_tokens(seg, skip_start, n)
            if target_off is None:
                body.append(_op_to_stmt(op, strings, items) or Stmt.raw_op(op.op, op.args))
                i += 1
                pending_cond = None
                continue
            skip_idx = _index_at_offset(ops, target_off)
            skip_idx = min(skip_idx, end)

            if pending_cond is None:
                pending_cond = Expr("unknown", {})

            if op.op == 0x10:
                else_body = lift_range(ops, seg, i + 1, skip_idx, strings, items)
                then_body = lift_range(ops, seg, skip_idx, end, strings, items)
            else:
                then_body = lift_range(ops, seg, i + 1, skip_idx, strings, items)
                else_body = lift_range(ops, seg, skip_idx, end, strings, items)

            body.append(Stmt.if_stmt(pending_cond, then_body, else_body))
            return body

        if op.op == 0x2B and op.args and pending_cond is None:
            n = op.args[0]
            skip_start = op.off + 2
            target_off = skip_n_tokens(seg, skip_start, n)
            if target_off is not None:
                skip_idx = min(_index_at_offset(ops, target_off), end)
                skipped = lift_range(ops, seg, i + 1, skip_idx, strings, items)
                rest = lift_range(ops, seg, skip_idx, end, strings, items)
                if skipped and rest:
                    body.extend(skipped)
                    body.extend(rest)
                    return body

        stmt = _op_to_stmt(op, strings, items)
        if stmt:
            body.append(stmt)
        pending_cond = None
        i += 1

    return body


def lift_segment(
    ops: list[LoweredOp],
    seg: bytes,
    strings: list[str],
    items: list[str],
) -> list[Stmt]:
    if not ops:
        return []
    try:
        lifted = lift_range(ops, seg, 0, len(ops), strings, items)
        return lifted if lifted else [_op_to_stmt(o, strings, items) or Stmt.raw_op(o.op, o.args) for o in ops]
    except Exception:
        fallback = [_op_to_stmt(o, strings, items) or Stmt.raw_op(o.op, o.args) for o in ops]
        return [Stmt.unlifted(fallback)]

File: editor/src/eventlang/CfgLift.cpp
#include "eventlang/CfgLift.h"

#include "eventlang/OpcodeTable.h"
#include "eventlang/Semantics.h"

#include <optional>

namespace mm2::eventlang {
namespace {

Expr opToCondExpr(const LoweredOp& op) {
    const uint8_t o = op.op;
    const auto& a = op.args;
    if (o == 0x09) return Expr::make("yes_no").set("mode", 0);
    if (o == 0x0A) return Expr::make("yes_no").set("mode", 1);
    // OP_16 @ 0x16520: arg1 discarded; only arg2 (item id) is scanned.
    if (o == 0x16 && a.size() >= 2)
        return Expr::make("party_has_item").set("a", a[0]).set("b", a[1]).set("item", a[1]);
    if (o == 0x1B && !a.empty()) return Expr::make("threshold").set("value", a[0]);
    // OP_1C @ 0x16742: cond = rng_roll(1..hi) raw byte.
    if (o == 0x1C && !a.empty()) return Expr::make("rng_roll").set("max", a[0]);
    if (o == 0x22 && a.size() >= 2) return Expr::make("era_in").set("lo", a[0]).set("hi", a[1]);
    if (o == 0x23 && a.size() >= 2) {
        if (a[0] == 0xB5) return Expr::make("day_odd");
        if (a[0] == 0xB6) return Expr::make("day_even");
        return Expr::make("day_in").set("lo", a[0]).set("hi", a[1]);
    }
    if (o == 0x24) {
        int v = decodeU16Arg(o, a.data(), static_cast<int>(a.size()));
        return Expr::make("gold_at_least").set("amount", v);
    }
    // OP_25 @ 0x16B82: pay gems (BE u16), NOT a "code" check.
    if (o == 0x25) {
        int v = decodeU16Arg(o, a.data(), static_cast<int>(a.size()));
        return Expr::make("gems_at_least").set("amount", v);
    }
    // OP_28 @ 0x16C86: consume backpack item → cond.
    if (o == 0x28 && a.size() >= 2)
        return Expr::make("consume_item").set("item", a[1]).set("probe", a[0]);
    if (o == 0x2D && a.size() >= 2)
        return Expr::make("member_attr").set("arg1", a[0]).set("arg2", a[1]);
    if (o == 0x17 && a.size() >= 2)
        return Expr::make("load_var8").set("group", a[0]).set("index", a[1]);
    if (o == 0x32 && !a.empty())
        return Expr::make("count_title_nibble").set("id", a[0]);
    if (o == 0x15 && a.size() >= 3) {
        return Expr::make("party_bits")
            .set("member", static_cast<int>(a[0]))
            .set("field", static_cast<int>(a[1]))
            .set("mask", static_cast<int>(a[2]));
    }
    // OP_19 sets cond=1 when item placed on a member.
    if (o == 0x19 && a.size() >= 4)
        return Expr::make("give_item_ok")
            .set("member", a[0])
            .set("item", a[1])
            .set("charges", a[2])
            .set("flags", a[3]);
    // OP_1F/20 @ 0x1690E: party field add/sub; cond=1 entry, OP_20 clears on underflow.
    if ((o == 0x1F || o == 0x20) && a.size() >= 6) {
        std::vector<int> args;
        for (uint8_t x : a) args.push_back(x);
        return Expr::make("party_effect_ok")
            .set("mode", o == 0x20 ? 1 : 0)
            .set("sel", a[0])
            .setList("args", args);
    }
    if (o == 0x30) {
        std::string decoded;
        for (uint8_t b : a) {
            int ch = (0x11A - b) & 0xFF;
            if (ch >= 0x20 && ch <= 0x7E) decoded += static_cast<char>(ch);
            else decoded += '?';
        }
        while (!decoded.empty() && decoded.back() == ' ') decoded.pop_back();
        return Expr::make("answer_eq").set("text", decoded);
    }
    Expr e = Expr::make("raw_op").set("op", o);
    std::vector<int> args;
    for (uint8_t x : a) args.push_back(x);
    e.setList("args", args);
    return e;
}

std::optional<Stmt> opToStmt(const LoweredOp& op) {
    const uint8_t o = op.op;
    const auto& a = op.args;
    // Cond-set ops: linear lift emits set_cond / ask_yes_no / side-effect stmts below.
    if (isCondSetOp(o) && o != 0x15 && o != 0x19 && o != 0x1F && o != 0x20 && o != 0x09 &&
        o != 0x0A && o != 0x17)
        return std::nullopt;
    if (o == 0x10 || o == 0x11 || o == 0x2B) return std::nullopt;
    if (o >= 0x01 && o <= 0x06 && !a.empty()) {
        return Stmt::make("say").set("variant", sayVariantForOp(o)).set("string", static_cast<int>(a[0]));
    }
    if (o == 0x07) return Stmt::make("wait").set("kind", "space").set("mode", 0);
    // OP_08 @ 0x15D26: still a SPACE wait; sets SCRIPTED_KEY_MODE=$FD first.
    if (o == 0x08) return Stmt::make("wait").set("kind", "space").set("mode", 1);
    if (o == 0x09) return Stmt::make("ask_yes_no").set("mode", 0);
    if (o == 0x0A) return Stmt::make("ask_yes_no").set("mode", 1);
    if (o == 0x0B && a.size() >= 2)
        return Stmt::make("service_title").set("sign", static_cast<int>(a[0])).set("mode", a[1]);
    if (o == 0x0C && a.size() >= 2)
        return Stmt::make("go_to").set("screen", a[0]).set("pos", a[1]);
    // OP_0D → play_sound_seq (ids 0..9), not a generic "engine" trap.
    if (o == 0x0D && !a.empty()) return Stmt::make("play_sound").set("id", a[0]);
    if (o == 0x0E && !a.empty()) return Stmt::make("selector").set("value", a[0]);
    if (o == 0x0F) return Stmt::make("end");
    if (o == 0x12 && a.size() >= 12) {
        std::vector<int> m, f;
        for (int i = 0; i < 10; ++i) m.push_back(a[i]);
        f.push_back(a[10]);
        f.push_back(a[11]);
        return Stmt::make("fight").setList("monsters", m).setList("flags", f);
    }
    if (o == 0x13 && a.size() >= 10) {
        std::vector<int> d;
        for (uint8_t x : a) d.push_back(x);
        return Stmt::make("fight_b").setList("data", d);
    }
    if (o == 0x14) return Stmt::make("clear_tile_event");
    if (o == 0x15 && a.size() >= 3)
        return Stmt::make("apply_party").set("count", a[0]).set("op", a[1]).set("val", a[2]);
    if (o == 0x18 && a.size() >= 4)
        return Stmt::make("set_party_bits")
            .set("member", a[0])
            .set("field", a[1])
            .set("and", a[2])
            .set("or", a[3]);
    if (o == 0x17 && a.size() >= 2)
        return Stmt::make("load_var8").set("group", a[0]).set("index", a[1]);
    if (o == 0x1A && a.size() >= 2)
        return Stmt::make("store_var8").set("group", a[0]).set("value", a[1]);
    if (o == 0x1D && !a.empty()) return Stmt::make("audio_wait").set("index", a[0]);
    if (o == 0x1E && !a.empty()) return Stmt::make("delay").set("ticks", a[0]);
    if ((o == 0x1F || o == 0x20) && a.size() >= 6) {
        std::vector<int> args;
        for (uint8_t x : a) args.push_back(x);
        return Stmt::make("party_effect")
            .set("mode", o == 0x20 ? 1 : 0)
            .set("sel", a[0])
            .setList("args", args);
    }
    if (o == 0x21 && a.size() >= 3) {
        int y = (a[0] >> 4) & 0xF, x = a[0] & 0xF;
        return Stmt::make("set_tile").set("y", y).set("x", x).set("a", a[1]).set("b", a[2]);
    }
    if (o == 0x26 || o == 0x27)
        return Stmt::make("select_member").set("mode", o == 0x27 ? 1 : 0);
    if (o == 0x19 && a.size() >= 4)
        return Stmt::make("give_item")
            .set("member", a[0])
            .set("item", a[1])
            .set("charges", a[2])
            .set("flags", a[3]);
    if (o == 0x29) return Stmt::make("abort");
    if (o == 0x2A && a.size() >= 14) {
        std::vector<int> d;
        for (uint8_t x : a) d.push_back(x);
        return Stmt::make("treasure").setList("data", d);
    }
    if (o == 0x2C && !a.empty()) return Stmt::make("add_counter").set("amount", a[0]);
    if (o == 0x2E && a.size() >= 2)
        return Stmt::make("or_member_field").set("arg1", a[0]).set("arg2", a[1]);
    if (o == 0x2F) return Stmt::make("read_answer");
    if (o == 0x31 && a.size() >= 3) {
        int v = a[1] | (a[2] << 8);
        return Stmt::make("party_damage").set("member", a[0]).set("value", v);
    }

    Stmt s = Stmt::make("raw_op").set("op", o);
    std::vector<int> args;
    for (uint8_t x : a) args.push_back(x);
    s.setList("args", args);
    return s;
}

// Fidelity-first linear lift: preserve exact OP_10/11/2B skip counts.
// Structured if/else is lossy for shared-tail / long-skip CFG patterns.
Stmt makeSkip(const char* kind, int n) {
    return Stmt::make(kind).set("n", n);
}

}  // namespace

std::vector<Stmt> liftSegment(const std::vector<LoweredOp>& ops, const uint8_t* /*seg*/,
                              size_t /*segLen*/) {
    std::vector<Stmt> body;
    body.reserve(ops.size());
    for (const auto& op : ops) {
        if (op.op == 0x10 && !op.args.empty()) {
            body.push_back(makeSkip("skip_if_true", op.args[0]));
            continue;
        }
        if (op.op == 0x11 && !op.args.empty()) {
            body.push_back(makeSkip("skip_if_false", op.args[0]));
            continue;
        }
        if (op.op == 0x2B && !op.args.empty()) {
            body.push_back(makeSkip("skip_if_victory", op.args[0]));
            continue;
        }

        // Pure cond-set ops → set_cond <expr> (sticky VM flag for following skips).
        // ask_yes_no / load_var8 / side-effecting cond ops have stmt forms.
        if (isCondSetOp(op.op) && op.op != 0x15 && op.op != 0x19 && op.op != 0x1F &&
            op.op != 0x20 && op.op != 0x09 && op.op != 0x0A && op.op != 0x17) {
            Stmt s = Stmt::make("set_cond");
            s.cond = opToCondExpr(op);
            body.push_back(std::move(s));
            continue;
        }

        auto st = opToStmt(op);
        if (st) {
            body.push_back(*st);
        } else {
            Stmt s = Stmt::make("raw_op").set("op", op.op);
            std::vector<int> args;
            for (uint8_t x : op.args) args.push_back(x);
            s.setList("args", args);
            body.push_back(s);
        }
    }
    return body;
}

}  // namespace mm2::eventlang
